File: src/329_easy_IsSubsequence.cpp
class Solution {
public:
    bool isSubsequence(string s, string t) {
        int p1=0, p2=0;
        while (p1 < s.size()) {
            while (p2 < t.size()) {
                if (s[p1] == t[p2]) break;
                p2 ++;
            }
            if (p2 == t.size()) return false;
            p1 ++;
            p2 ++;
        }
        return p1 == s.size();
    }
};

File: src/329_easy_IsSubsequence.py
class Solution:
    def isSubsequence(self, s: str, t: str) -> bool:
        ch2nums = defaultdict(list)
        for i,ch in enumerate(t):
            ch2nums[ch].append(i)
        
        records = []
        for ch in s:
            if ch not in ch2nums:
                return False
            
            if len(records) > 0:
                while ch2nums[ch][0] <= records[-1]:
                    del ch2nums[ch][0]
                    if len(ch2nums[ch]) == 0:
                        return False 

            records.append(ch2nums[ch][0])
            del ch2nums[ch][0]
            if len(ch2nums[ch]) == 0:
                del ch2nums[ch]
        
        return True